File: codejam/16/0q/d-fractiles.py
#!/usr/bin/env python3

def geometric_series(base, power):
    if base == 1:
        return 1
    return (base**power - 1) // (base - 1)


def positions(pattern, complexity):
    base_position = geometric_series(pattern, complexity)
    return [placement*base_position for placement in range(pattern)]


def main():
    for case in range(int(input())):
        pattern, complexity, _ = [int(n) for n in input().split()]
        answer = (pos+1 for pos in positions(pattern, complexity))
        print('Case #{}: {}'.format(case+1, ' '.join(str(n) for n in answer)))


if __name__ == '__main__':
    main()

File: codejam/16/0q/d-fractiles.hs
import Text.Printf
import Data.List.Split (splitOn)


getInts :: IO [Int]
getInts = do
    it <- getLine
    return $ [read n | n <- splitOn " " it]


printList :: [Int] -> IO ()
printList [] = do printf "\n"
printList (g:gs) = do
    printf " %i" g
    printList gs


geometricSeries :: Int -> Int -> Int
geometricSeries 1    _      = 1
geometricSeries base power = quot (base^power - 1) (base - 1)


goldLocations :: Int -> Int -> [Int]
goldLocations pattern complexity =
    let basePosition = geometricSeries pattern complexity
    in  [placement * basePosition + 1 | placement <- [0..pattern-1]]


test :: Int -> IO ()
test t = do
    [pattern,complexity,_] <- getInts
    printf "Case #%i:" t
    printList $ goldLocations pattern complexity


main :: IO ()
main = do
    [loop] <- getInts
    sequence_ [test t | t <- [1..loop]]
